common/subtitle-coloring/subtitle-coloring.ts

```diff
@@ -0,0 +1,777 @@
+            this.colorCacheBuilding = false;
+        }
+        return !buildWasCancelled;
+    }
+
+    private async _colorizeText(options: { text: string; index: number; ts: TrackState }): Promise<string | undefined> {
+        const { text, index, ts } = options;
+        try {
+            if (!this.anki) throw new Error('Anki not initialized');
+            if (!ts.yomitanHealthy) throw new Error(`Yomitan not initialized for Track${ts.track + 1}`);
+            if (!ts.dt.dictionaryAnkiWordFields.length && !ts.dt.dictionaryAnkiSentenceFields.length) {
+                throw new Error('No Anki fields defined');
+            }
+
+            let richText: string = '';
+            let textHasError = false;
+            let textHasUncollected = false;
+            const rawTokens = await this.yomitan.tokenize(
+                ts.track,
+                text,
+                ts.dt.dictionaryYomitanScanLength,
+                ts.dt.dictionaryYomitanUrl
+            );
+            if (this.shouldCancelBuild) return;
+            for (const rawToken of rawTokens) {
+                const trimmedToken = rawToken.trim();
+
+                // Token is already cached or not a word
+                const cachedTokenStatus = ts.tokenStatusCache.get(trimmedToken);
+                if (this._tokenStatusValid(cachedTokenStatus)) {
+                    const richRes = this._applyTokenStyle({ rawToken, tokenStatus: cachedTokenStatus!, dt: ts.dt });
+                    if (richRes) richText += richRes;
+                    if (cachedTokenStatus === TokenStatus.UNCOLLECTED) textHasUncollected = true;
+                    else if (cachedTokenStatus === null) textHasError = true;
+                    continue;
+                }
+                if (!HAS_LETTER_REGEX.test(trimmedToken)) {
+                    const fullyKnownTokenStatus = getFullyKnownTokenStatus();
+                    const richRes = this._applyTokenStyle({ rawToken, tokenStatus: fullyKnownTokenStatus, dt: ts.dt });
+                    if (richRes) richText += richRes;
+                    ts.tokenStatusCache.set(trimmedToken, fullyKnownTokenStatus);
+                    continue;
+                }
+
+                let shouldCheckExactFormWordField = true;
+                if (ts.dt.dictionaryTokenMatchStrategy === TokenMatchStrategy.LEMMA_FORM_COLLECTED) {
+                    const tokenLemmas = await this.yomitan.lemmatize(
+                        ts.track,
+                        trimmedToken,
+                        ts.dt.dictionaryYomitanUrl
+                    );
+                    if (this.shouldCancelBuild) return;
+                    if (tokenLemmas.length) shouldCheckExactFormWordField = false;
+                }
+                let shouldCheckExactFormSentenceField = true;
+                if (ts.dt.dictionaryAnkiSentenceTokenMatchStrategy === TokenMatchStrategy.LEMMA_FORM_COLLECTED) {
+                    const tokenLemmas = await this.yomitan.lemmatize(
```

Comment:
Minor point, but if we really do need this block then I suggest merging it with the one above to reduce duplicated code. The only difference in logic is whether `shouldCheckExactFormWordField ` or `shouldCheckExactFormSentenceField` gets assigned.